hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h

```diff
@@ -336,7 +338,14 @@ class BaseSocket : public hazelcast::client::socket
     boost::asio::ip::tcp::resolver& resolver_;
     T socket_;
     int32_t call_id_counter_{ 0 };
-    typedef std::deque<std::vector<boost::asio::const_buffer>> Outbox;
+    struct entry
+    {
+        std::vector<boost::asio::const_buffer> buffers;
+        std::shared_ptr<spi::impl::ClientInvocation> invocation;
+        std::shared_ptr<protocol::ClientMessage> message;
```

Comment:
The `message` should be accessible via invocation, right? why keep it separate?

Reply:
because of the retry mechanism, the message might be deleted from invocation. That is why message is hold in this structure.